[Wallet] New exchange trade screen and fix ts build in celotool

annakaz opened the pull request:
### Description

Add new Exchange trade screen

Previously, users could toggle between Buying/Selling gold or dollars and entering the value in whichever currency they were going to trade in. The new UI trade screen defines a direction (Buy or Sell) before users input the amount they'd like to trade. This allows us to display a more accurate exchange rate, as we know which direction to show the exchange rate in -- there is small buy/sell spread in the Exchange rate (ex: to sell, receive $9.9499, to buy, pay $10.0502). 

This rate also varies (an even smaller amount than with direction) based on the size of the exchange (ex: selling 1 gold gives rate of $0.1005, 1000 gold a rate of $0.1006), so the 'worst case' rate is shown by estimating the rate at the maximum amount of gold/dollars a user could buy or sell. 

### Tested

Sold and bought gold on `alfajoresstaging`
Updated jest tests

### Other changes

Fix empty fee in confirmation screen
Added missingFullName error text

### Related issues

- Fixes #1728 
- Fixes #1396 (empty fee)
- Fixes #1438 , #1705 (bugs in old trade UI)

Note that local currency-related work is a separate issue (#1756 ) as it will build on top of https://github.com/celo-org/celo-monorepo/issues/1484 

### Backwards compatibility

Yes

jeanregisser committed:
Set default gradle properties

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/set-default-gradle-properties

jeanregisser committed:
Testing reduced jvm options since we're setting them in
gradle.properties

annakaz committed:
Merge branch 'jeanregisser/set-default-gradle-properties' into annakaz/gold-exchange

annakaz committed:
Fix Buy Sell buttons

annakaz committed:
Remove unused field component

annakaz committed:
Update fonts based on Corey spec

annakaz committed:
Add new exchange screen

annakaz committed:
WIP

annakaz committed:
WIP new gold buy screen

annakaz committed:
WIP

annakaz committed:
Fix passing balance to exchange screen

annakaz committed:
Logic working

annakaz committed:
WIP final styling

annakaz committed:
Finish screen ui

annakaz committed:
Clean up css

annakaz committed:
Organize css

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Update buy/sell gold screen

annakaz committed:
Finish buy/sell screen

annakaz committed:
Basic confirmation card layout

annakaz committed:
WIP on review screen

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Adding missing fullname error to language jsons

annakaz committed:
WIP exchange rate directions

annakaz committed:
Fix lint errors

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
WIP

annakaz committed:
Rename buy screen to trade screen

annakaz committed:
WIP found contractkit bug

annakaz committed:
Get exchange direction rate correct

annakaz committed:
Clean up exchange review

annakaz committed:
Remove logging statements

annakaz committed:
Clean up saga

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Remove unused state

annakaz committed:
Separate exchange header

annakaz committed:
Update translations

annakaz committed:
Update translations

annakaz committed:
Clean up css, fix conversion

annakaz committed:
Add fee icon, fix confirmation card missing fee

annakaz committed:
Use currency enum not token

annakaz committed:
Remove keyboard scrollview

annakaz committed:
Update convert to contract decimals

annakaz committed:
Add grey icon

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Update tests

annakaz committed:
PR clean up

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Fix lint

annakaz commented:
Screenshots: 
![Screenshot_20191210-154739](https://user-images.githubusercontent.com/16907027/70579194-b5dc1380-1b64-11ea-85ff-a922b5a08449.png)
![Screenshot_20191210-154746](https://user-images.githubusercontent.com/16907027/70579196-b7a5d700-1b64-11ea-92e5-5dd8b316ac3e.png)
![Screenshot_20191210-154752](https://user-images.githubusercontent.com/16907027/70579198-b96f9a80-1b64-11ea-8dd1-9cc48748947b.png)
![Screenshot_20191210-154755](https://user-images.githubusercontent.com/16907027/70579205-baa0c780-1b64-11ea-8053-a9562622f27c.png)
![Screenshot_20191210-154809](https://user-images.githubusercontent.com/16907027/70579211-bbd1f480-1b64-11ea-8886-7a0215cf54c4.png)
![Screenshot_20191210-154830](https://user-images.githubusercontent.com/16907027/70579214-be344e80-1b64-11ea-9f74-aa5093a3b166.png)

annakaz committed:
Fix alignment

annakaz committed:
Add tests

annakaz committed:
Add Review test

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz commented:
Note that build is failing on master so tests aren't running. `yarn run test` in mobile passes locally.

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Fee Icon work

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

jeanregisser commented on packages/mobile/src/exchange/ExchangeHeader.tsx:
```diff
@@ -0,0 +1,33 @@
+          <Text style={fontStyles.subSmall}>
+            {getMoneyDisplayValue(makerTokenBalance, makerToken, true) +
+              ' ' +
+              i18n.t(`${Namespaces.exchangeFlow9}:lowerCaseAvailable`)}
```
For better translation we should avoid concatenating strings this way and use interpolation:

```ts
t('moneyAvailable', { moneyAmount: getMoneyDisplayValue(makerTokenBalance, makerToken, true)})
```

And have the following in the locale json file:

```json
moneyAvailable: "{{moneyAmount}} available"
```

jeanregisser commented on packages/mobile/src/exchange/ExchangeHeader.tsx:
```diff
@@ -0,0 +1,33 @@
+  const title =
+    makerToken === CURRENCY_ENUM.DOLLAR
+      ? i18n.t(`${Namespaces.exchangeFlow9}:buyGold`)
+      : i18n.t(`${Namespaces.exchangeFlow9}:sellGold`)
```
We should inject `t` using `withNamespace` instead of directly accessing i18n, main reason behind is the new translations get re-rendered when switching language.

Note that once we can upgrade to the latest i18next version we'll be able to use a hook instead which is simpler IMHO.

annakaz replied:
I moved this exchangeHeader to Headers.tsx, so we can't inject t using withNamespace. I use the global i18n.t following what we do for other header titles (see Analytics.tsx or Invite.tsx). Let me know if you think another approach would be better

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -136,9 +232,33 @@ class ExchangeReview extends React.Component<Props> {
+    paddingHorizontal: 16,
+  },
+  column: {
+    flexDirection: 'column',
```
nit: this is the default direction

annakaz replied:
Thanks for catching this

jeanregisser commented on packages/mobile/src/exchange/ExchangeTradeScreen.test.tsx:
```diff
@@ -55,20 +66,48 @@ describe(ExchangeTradeScreen, () => {
+      expect(mockShowError).toBeCalledWith(ErrorMessages.NSF_DOLLARS) // Can't afford 20 cUSD worth of gold
     })
 
     it.only('validates amount', () => {
```
Looks like only this test case will run, do you know why we did this?

annakaz replied:
Removed, now we run them all

jeanregisser commented on packages/mobile/src/icons/InfoIcon.tsx:
```diff
@@ -3,17 +3,21 @@ import { Image } from 'react-native'
-        source={require('src/images/info.png')}
+        source={
+          this.props.isGrey ? require('src/images/info-grey.png') : require('src/images/info.png')
+        }
```
`<Image />` style supports `tintColor` which would be better here instead of adding a new image.
In that case it would be more flexible to replace `isGrey` by `tintColor` (either enum of string depending on the flexibility we want).

annakaz replied:
Good idea, fixed!

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,229 @@
+    makerToken: CURRENCY_ENUM
+    makerTokenBalance: string
+    inputToken: CURRENCY_ENUM
+    inputTokenCode: string
```
Why do we need `inputTokenCode` given we already have `inputToken`?

annakaz replied:
To avoid having to figure out the string again. In the future I would see us using CURRENCIES.code instead, but in the mean time I wanted to avoid copying over the `getInputTokenDisplayText` function

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,229 @@
+            <View style={styles.column}>
+              <View style={[styles.rowContainer, styles.amountRow]}>
+                <Text style={[fontStyles.body, styles.exchangeBodyText]}>
+                  {t('amount') + ` (${this.state.inputTokenCode})`}
```
As commented above, we should use a single translation key and use interpolation.

Approximate implementation ([using nesting and interpolation](https://www.i18next.com/translation-function/nesting#passing-nesting-to-interpolated), but there are other ways):

```ts
t('exchangeAmount', {tokenName: `$t('global:${ key }')` })
```

Locale json
```json
exchangeAmount: "Amount ({{tokenName}})"
```

annakaz replied:
Fixed

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,229 @@
+          <ScrollView>
+            <View style={styles.column}>
+              <View style={[styles.rowContainer, styles.amountRow]}>
+                <Text style={[fontStyles.body, styles.exchangeBodyText]}>
```
Please pass a single style without an array when there's no dynamic part. It prevents re-rendering unnecessarily.

```ts
exchangeBody: {
  ...fontStyles.body,
  xxx
}
```

Same applies to the other styles used in this screen.

annakaz replied:
Fixed!

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,229 @@
+                <Text style={[fontStyles.body, styles.exchangeBodyText]}>
+                  {t('subtotal') +
+                    ' @ ' +
+                    getMoneyDisplayValue(exchangeRate, CURRENCY_ENUM.DOLLAR, true)}
```
Same comment about using interpolation instead of concatenation which can't be translated correctly.

jeanregisser commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,229 @@
+              ) +
+              ' ' +
+              t('global:gold')
+            }
```
Here as well, please use interpolation.

jeanregisser reviewed (changes requested):
Awesome work! 👍 🎉 

Comments are mostly about our coding style and a better way of doing localization.

annakaz committed:
Merge branch 'annakaz/gold-exchange' of github.com:celo-org/celo-monorepo into annakaz/gold-exchange

annakaz committed:
Use interpolation, update header, fix styles

annakaz committed:
Remove logs, fix FeeExchange navigator

annakaz commented:
Thanks Jean! Switched to using interpolation for translation and a single style where applicable. Also moved the exchangeHeader to Headers.tsx, so using the global i18n variable for translations there.

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz commented on packages/mobile/src/navigator/Navigator.tsx:
```diff
@@ -296,6 +296,7 @@ const AppStack = createStackNavigator(
     [Screens.PhotosEducation]: { screen: PhotosEducation },
     [Screens.GoldEducation]: { screen: GoldEducation },
     [Screens.FeeEducation]: { screen: FeeEducation },
+    [Screens.FeeExchangeEducation]: { screen: FeeExchangeEducation },
```
Added so we navigate back to exchange confirmation card, rather than always going to exchange. 
Ex flow this fixes: Click exchange notification from home screen, then click FeeExchangeEducation, expect to return back to exchange confirmation card but instead go to the exchange

jmrossy replied:
Maybe leave a comment explaining this?

annakaz committed:
PR cleanup

annakaz committed:
Restore default env

annakaz committed:
Merge branch 'annakaz/gold-exchange' of github.com:celo-org/celo-monorepo into annakaz/gold-exchange

jeanregisser reviewed (approved):
Looking great! 💯

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Update tests

Commit:
Merge refs/heads/master into annakaz/gold-exchange

Commit:
Merge refs/heads/master into annakaz/gold-exchange

Commit:
Merge refs/heads/master into annakaz/gold-exchange

Commit:
Merge refs/heads/master into annakaz/gold-exchange

annakaz committed:
Add space to trigger build

annakaz committed:
Remove space added to trigger build

jmrossy commented on packages/mobile/locales/en-US/nuxVerification2.json:
```diff
@@ -1,6 +1,7 @@
 {
   "getVerified": "Verify Phone Number",
   "notReadyForCode": "Not yet ready for code entry",
+  "missingFullName": "Missing Full Name",
```
This was moved to global.json, please remove :)

jmrossy commented on packages/mobile/locales/es-419/nuxVerification2.json:
```diff
@@ -1,6 +1,7 @@
 {
   "getVerified": "Verificar el Teléfono",
   "notReadyForCode": "Aún no está listo para ingresar el código",
+  "missingFullName": "Nombre completo perdido",
```
Ditto

jmrossy commented on packages/mobile/src/components/FeeIcon.tsx:
```diff
@@ -0,0 +1,43 @@
+    navigate(Screens.FeeEducation, {})
+  }
+
+  navigateToExchangeEducate() {
```
this will work fine in this case but FYI, handlers should generally be lamdbas instead of class methods so that the `this` keyword gets bound properly

annakaz replied:
Kk good to know!

jmrossy commented on packages/mobile/src/exchange/ExchangeConfirmationCard.tsx:
```diff
@@ -130,15 +129,15 @@ export function ExchangeConfirmationCard(props: Props) {
         <LineItemRow
           currencySymbol={takerToken}
-          amount={fee}
+          amount={fee || '0.00'}
```
would undefined be a better choice here to signify that the fee isn't available yet? Nbd either way

annakaz replied:
Thanks for flagging this- the ExchangeConfirmationCard is now only used for reviewing past exchange (so we should not have to wait to fetch the fee), but the current fee and exchangeRate are being fetched instead of using the historic ones. Have created an issue here to track: https://github.com/celo-org/celo-monorepo/issues/2269

jmrossy commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,224 @@
+  state: State = {
+    makerToken: CURRENCY_ENUM.GOLD,
+    inputToken: CURRENCY_ENUM.GOLD,
+    inputTokenCode: this.props.t('global:gold'),
```
Is the tokenCode a value used for display purposes? If so, maybe rename to something more clear. `inputTokenDisplayName`?

jmrossy replied:
Also, though it's probably fine in this case, using props in state initialization like this is generally considered an anti-pattern:
https://medium.com/@justintulk/react-anti-patterns-props-in-initial-state-28687846cc2e

jmrossy commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -1,133 +1,224 @@
+                  {getMoneyDisplayValue(this.state.inputAmount, this.state.inputToken, true)}
+                </Text>
+              </View>
+              <View style={styles.line} />
```
Can we use `HorizontalLine` from react-components here?

jmrossy commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -136,8 +227,38 @@ class ExchangeReview extends React.Component<Props> {
     flex: 1,
-    backgroundColor: colors.background,
-    paddingTop: 20,
+    flexDirection: 'column',
```
column is the default

jmrossy commented on packages/mobile/src/exchange/ExchangeReview.tsx:
```diff
@@ -136,8 +227,38 @@ class ExchangeReview extends React.Component<Props> {
+  paddedContainer: {
+    paddingHorizontal: 16,
+  },
+  column: {
```
Nit: strange name for something that just justifies to flex-start

annakaz replied:
Good point. Renamed to flexStart

jmrossy commented on packages/react-components/styles/fonts.tsx:
```diff
@@ -167,6 +167,11 @@ export const fontStyles = StyleSheet.create({
     fontFamily: HindSilguri.Bold,
     color: colors.dark,
   },
+  headerBoldTitle: {
```
This is the same as `headerTitle` above it but with one font-size more. Can we merge these two? If not, consider renaming this.

annakaz replied:
Good point- will merge these to one

jmrossy reviewed (changes requested):
Overall really nice work! Just some minor things to address

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Make changes based on review

annakaz committed:
Update tests

jmrossy reviewed (approved):
Nice!

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Remove console.log

annakaz committed:
Merge branch 'annakaz/gold-exchange' of github.com:celo-org/celo-monorepo into annakaz/gold-exchange

annakaz committed:
Merge branch 'master' into annakaz/gold-exchange

annakaz committed:
Add missing spanish translation

annakaz committed:
Add space to trigger build

annakaz committed:
Remove space to trigger build

annakaz committed:
Trigger build

annakaz committed:
Add space to trigger build

annakaz committed:
Remove space to trigger build

annakaz committed:
Fix build errors from ts upgrade
